src/wallet/wallet2.cpp

```diff
@@ -5170,15 +5173,35 @@ bool wallet2::check_connection(uint32_t *version, bool *ssl, uint32_t timeout)
       return m_light_wallet_connected;
   }
 
+  bool rpc_failed = false, reconnection_made = false;
+  MTRACE("First attempt");
+again:
```

Comment:
I'm not 100% sure, but I think it's the ssl context that needs resetting.

Reply:
The context should be reusable, but that may have been botched.

Can this be solved by moving the SSL context initialization from `set_ssl` to `connect` in `epee/net/net_helper.h` ? This would solve this issue across all SSL HTTP usages. Also, what is the issue/how would I reproduce (if help is necessary)?

Reply:
I wanted to avoid doing so because creating the context is substantially slower, so if it's not needed I'd rather not do it. However, it might be that in practice the case never happens where we reconnect without the need for reset ?
Anyway, the test case is to start monerod and monero-wallet-cli, then restart monerod. With the current master code, monero-wallet-cli is unable to reconnect to the restarted daemon.

Reply:
The issue is not related to SSL. What happens is that the existing connection is dead, so the invoke command fails but puts the HTTP client into the disconnected state. The "critical" part with this patch is the re-connect attempt.

The easiest way to see whats happening - run the master branch with or without SSL. Restart `monerod` after `monero-wallet-cli` connects (i.e. the test case you just suggested). Calling `refresh` in `monero-wallet-cli` will print an error message, then running `refresh` a second time will succeed. And this happens with SSL enabled or disabled on master.

So the only thing that is needed is a retry attempt _somewhere_ after a send fails with a lost connection `error_code` (whatever its called). Doing it here catches a few specific cases but not all. Might be worth figuring how to make this work in either the HTTP client or the `invoke` helper function.

Either way, just purge the SSL caching stuff because it doesn't appear to be needed, its not related.

Reply:
With autodetect SSL, this case had monero-wallet-cli *never* manage to reconnect unless I reset that SSL state. But maybe that's part of the 1.70.0 beta boost thing. I'll try with a final 1.70.0.

Reply:
I updated to 1.70.0, and I applied this:

```
diff --git a/src/wallet/wallet2.cpp b/src/wallet/wallet2.cpp
index d4b0fadf6..bb709ee69 100644
--- a/src/wallet/wallet2.cpp
+++ b/src/wallet/wallet2.cpp
@@ -5263,11 +5263,13 @@ again:
       m_node_rpc_proxy.invalidate();
       if (!m_http_client.connect(std::chrono::milliseconds(timeout)))
       {
+/*
         if (!m_http_client.set_server(get_daemon_address(), get_daemon_login(), m_ssl_options))
         {
           MERROR("set_server failed");
           return false;
         }
+*/
         if (!m_http_client.connect(std::chrono::milliseconds(timeout)))
         {
           MERROR("reconnection failed");
```

This makes the problem reappear. This reconnects all the time, not only on any particular error code, but that should not matter. After the daemon restarts, the wallet cannot reconnect.

Reply:
Ok, so it was as I first thought -

The `autodetect` setting gets overwritten to `enabled` in `epee/net/net_helper.h`, which changes the verification behavior of the next handshake. That class should probably remain in (1) `autodetect` mode or (2) set `enabled` + cache cert fingerprint for subsequent connections. Calling `set_server` is resetting to `autodetect` (1), but only for the wallet and with the downside that the SSL context is being re-generated, etc.

Leaving the state in `autodetect` mode (1) means a mitm could occur on reconnect, whereas current behavior requires a CA signature to perform a mitm on reconnect. Switching to `enabled` mode with cached fingerprint (2) will reject valid CA signed connections in rare cases, but cannot be mitm after the initial connection (and until the process exits).

Reply:
What is the case that could be rejected in 2 ? If the daemon restarts with a new cert, it's very unlikely that the wallet had whitelisted the previous cert since it's most likely set to random. Another case ?

Reply:
I'm confused by 2. For default random daemon cert, caching the cert will not help.  If you set ssl to enabled in the first place, it should connect fine, theoretically (if MITM friendly). So if the daemon restarts and the wallet switched to enabled, I don't see what is wrong in theory.

Reply:
I probably added too much confusion to this.

The client in `epee/net/net_helper.h` is setting SSL to `enabled` if `autodetect` succeeds causing verification to fail on next connect (autodetect does not fail if SSL cert is not signed). Leaving the client (in `net_helper.h`) in `autodetect` mode does the same thing without resetting the SSL context (which this patch is doing). Changing in `net_helper.h` also fixes other cases. It has one downside: if a server got their cert CA signed, `autodetect` -> `enabled` provides slightly better checking after first connection.

Also, `net_helper.h` requires a few other edits to make the `autodetect` change work but they aren't hard. I was going to include them in my [recent PR](https://github.com/monero-project/monero/pull/5561), but didn't because it was unrelated to what I was trying to fix for that PR.

Reply:
I'm going to leave the patch as is for now since it works. We can change it later if/when I understand your point.

Reply:
Do you see actual bugs with the patch ? If not, I'll also PR to the branch.

Reply:
The bug is [actually here](https://github.com/monero-project/monero/blob/9956500d142349edae4d1f2692674382b04b1ac6/contrib/epee/include/net/net_helper.h#L226) and the issue is present with all SSL client usage. This patch fixes the wallet usage in one specific code path.

As for comments on this code directly - a loop with a counting conditional would be preferable to a `goto`. It is more readable and an infinite loop case is prevented. The easy "fix" for this patch is to add `reconnection_made = true` just before the `goto` on line 5218. If someone pointed this to a HTTP server that wasn't `monerod` by accident, this should/could loop indefinitely - the connection on line 5184 succeeds and then the invoke fails on line 5211.

Reply:
OK, removing this works, but I think that's going to have side effects, which I'm going to look for now.

Reply:
I had tested without removing the set_daemon call. It does not work after removing it.

Reply:
I don't understand what you removed. The `= enabled` line in `net_helper.h`? There were a few other mods that were necessary in that file unfortunately. I would recommend adding `reconnection_mode = true` before `goto` on line 5218 and that's it. I've _almost_ got a sensible change to the SSL/HTTP client classes that should handle this and other cases too, but it won't be ready for a bit. Hopefully it will make more sense after the code changes are shown.

Reply:
I removed (1) the line you pointed to, which sets the ssl mode to enabled, and (2) the set_daemon call in wallet2.cpp, which you reckon should not be needed.

Reply:
I mentioned in passing (in this very long stream) that multiple edits in `net_helper.h` are required. I just made the relevant changes to verify, and it is indeed that setting of `enabled` in  that file (its done in two places). I'll post something later hopefully, but its going to a slightly larger refactor so that the behavior is better when the server has a CA signed cert.